File: python/251_flatten_2D_vector.py
class Vector2D:

    def __init__(self, v: List[List[int]]):
        def it():
            for line in v:
                for val in line:
                    self.size -= 1
                    yield val
        self.it = it()
        self.size = sum(len(line) for line in v)

    def next(self) -> int:
        return next(self.it)

    def hasNext(self) -> bool:
        return self.size > 0


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(v)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: 251_flatten_2d_vector.java
public class Vector2D implements Iterator<Integer> {

    Queue<Integer> queue;

    public Vector2D(List<List<Integer>> vec2d) {
        queue = new LinkedList<>();
        for (List<Integer> list: vec2d) {
            for (Integer num: list) {
                queue.offer(num);
            }
        }
    }

    @Override
    public Integer next() {
        return queue.poll();
    }

    @Override
    public boolean hasNext() {
        return ! queue.isEmpty();
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i = new Vector2D(vec2d);
 * while (i.hasNext()) v[f()] = i.next();
 */
